src/core/server.cc

```diff
@@ -294,150 +293,354 @@ class HTTPServiceImpl {
   {
   }
 
-  tfs::net_http::RequestHandler Dispatch(
-      tfs::net_http::ServerRequestInterface* req);
+  static void Dispatch(evhtp_request_t* req, void* arg);
+
+  tensorflow::Status Start(uint16_t port, int thread_cnt) override;
+
+  tensorflow::Status Stop() override;
 
  private:
-  void Handle(tfs::net_http::ServerRequestInterface* req);
-  tfs::net_http::HTTPStatusCode Health(
-      tfs::net_http::ServerRequestInterface* req,
-      const std::string& health_uri);
-  tfs::net_http::HTTPStatusCode Profile(
-      tfs::net_http::ServerRequestInterface* req,
-      const std::string& profile_uri);
-  tfs::net_http::HTTPStatusCode Infer(
-      tfs::net_http::ServerRequestInterface* req, const std::string& infer_uri);
-  tfs::net_http::HTTPStatusCode Status(
-      tfs::net_http::ServerRequestInterface* req,
-      const std::string& status_uri);
+  void Handle(evhtp_request_t* req);
+  void Health(evhtp_request_t* req, const std::string& health_uri);
+  void Profile(evhtp_request_t* req, const std::string& profile_uri);
+  void Infer(evhtp_request_t* req, const std::string& infer_uri);
+  void Status(evhtp_request_t* req, const std::string& status_uri);
+  static std::string ToMethodString(htp_method method);
+
+  static void thread_init(evhtp_t* htp, evthr_t* thread, void* arg);
+  static void thread_exit(evhtp_t* htp, evthr_t* thread, void* arg);
+
+  // class object associated to evhtp thread, requests received are bounded
+  // with the thread that accepts it. Need to keep track of that and let the
+  // corresponding thread send back reply
+  class ThreadWiseService {
+   public:
+    class Request {
+     public:
+      Request(
+          evhtp_request_t* req, uint64_t id,
+          std::shared_ptr<HTTPInferRequestProvider> request_provider,
+          std::shared_ptr<HTTPInferResponseProvider> response_provider)
+          : done(false), req(req), id(id), request_provider(request_provider),
+            response_provider(response_provider){};
+
+     private:
+      friend class HTTPServiceImpl;
+      friend class ThreadWiseService;
+      bool done;
+      evhtp_request_t* req;
+      uint64_t id;
+      RequestStatus request_status;
+      std::shared_ptr<HTTPInferRequestProvider> request_provider;
+      std::shared_ptr<HTTPInferResponseProvider> response_provider;
+    };
+
+    ThreadWiseService(evthr_t* thread) : thread_(thread){};
+    ~ThreadWiseService() = default;
+
+    static void ReplyCallback(evthr_t* thr, void* arg, void* shared)
+    {
+      ThreadWiseService* requests = (ThreadWiseService*)arg;
+      Request* request = nullptr;
+      {
+        std::lock_guard<std::mutex> lock(requests->mtx_);
+        // Using iterator will cause segfault
+        for (size_t i = 0; i < requests->requests_.size(); i++) {
```

Comment:
I don't understand this loop. Why aren't you just doing the response for one request? If there are multiple done requests you just seem to erase them all and only remember the last one. Seems like there could be lost/leaked references. Using erase this way can be very inefficient.

Why do you even need requests_ at all? Doesn't the callback return the Request object?

Reply:
I didn't implement it correctly, one invocation of the callback should only send and erase one selected request. I am going to fix it like you said, directly passing the Request object to the callback.